log4j-api/src/main/java/org/apache/logging/log4j/ScopedContext.java

```diff
@@ -0,0 +1,224 @@
+import java.util.concurrent.Callable;
+import java.util.function.Supplier;
+
+/**
+ * Context that can be used for data to be logged in a block of code.
+ *
+ * While this is influenced by ScopedValues from Java 21 it does not share the same API. While it can perform a
+ * similar function as a set of ScopedValues it is really meant to allow a block of code to include a set of keys and
+ * values in all the log events within that block. The underlying implementation must provide support for
+ * logging the ScopedContext for that to happen.
+ *
+ * The ScopedContext will not be bound to the current thread until either a run or call method is invoked. The
+ * contexts are nested so creating and running or calling via a second ScopedContext will result in the first
+ * ScopedContext being hidden until the call is returned. Thus the values from the first ScopedContext need to
+ * be added to the second to be included.
+ *
+ * @since 2.24.0
+ */
+public class ScopedContext {
+
+    private static final ThreadLocal<Deque<Map<String, Renderable>>> scopedContext = new ThreadLocal<>();
+
+    /**
+     * Returns an immutable Map containing all the key/value pairs as Renderable objects.
+     * @return An immutable copy of the Map at the current scope.
+     */
+    public static Map<String, Renderable> getContext() {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack != null && !stack.isEmpty()) {
+            return Collections.unmodifiableMap(stack.getFirst());
+        }
+        return Collections.emptyMap();
+    }
+
+    private static void addScopedContext(Map<String, Renderable> contextMap) {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack == null) {
+            stack = new ArrayDeque<>();
+            scopedContext.set(stack);
+        }
+        stack.addFirst(contextMap);
+    }
+
+    private static void removeScopedContext() {
+        Deque<Map<String, Renderable>> stack = scopedContext.get();
+        if (stack != null) {
+            if (!stack.isEmpty()) {
+                stack.removeFirst();
+            }
+            if (stack.isEmpty()) {
+                scopedContext.remove();
+            }
+        }
+    }
+
+    /**
+     * Return a new ScopedContext.
+     * @return the ScopedContext.
+     */
+    public static ScopedContext newInstance() {
+        return newInstance(false);
+    }
+
+    /**
+     * Return a new ScopedContext.
+     * @param inherit true if this context should inherit the values of its parent.
+     * @return the ScopedContext.
+     */
+    public static ScopedContext newInstance(boolean inherit) {
+        return new ScopedContext(inherit);
+    }
+
+    /**
+     * Return the key from the current ScopedContext, if there is one and the key exists.
+     * @param key The key.
+     * @return The value of the key in the current ScopedContext.
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> T get(String key) {
```

Comment:
I believe that `ThreadContext` and `ScopedContext` should be **write-only**.

As far as I see it, these are logging-only services. The presence of `get()` methods allows user to use these services to store and retrieve their thread-dependent data, which is in my opinion and **abuse** of the API.

There are more proper ways to store data, e.g. in a servlet application the username should be stored in the `ServletRequest` attributes, not in `ThreadContext`.

Reply:
I have to disagree with this but here we are just expressing our own opinions. Your example would require passing the ServletRequest through many methods that have no business knowing what a ServletRequest is or that they are even running in a servlet, but do need to know who the user is. For example, I have plenty of code that is shared between REST requests and Kafka consumers.  It is obviously very late to suggest that the ThreadContext be write only.   FWIW, the enhancement going into SLF4J does something sort of similar to this but instead of having the scoped context completely separate there you still do MDC.get to fetch attributes from the scoped context.

Reply:
Yes, I believe it is a matter of taste and commodity.

I am a little bit worried about users using `ThreadContext` for business-critical functionality, since a change of logging backend (e.g. from `log4j-core` to `log4j-to-jul`), might break that functionality. For example in #2374 I assumed that the `ThreadContextMap` implementation used by an implementation is strictly linked to the logger context implementation; since `java.util.logging` formatters don't support context data, `log4j-to-jul` might as well use no-op `ThreadContextMap`. Am I wrong?